File: 2021/08/10/ContainsNearbyDuplicate.java
/**
 * https://leetcode.com/problems/contains-duplicate-ii/
 * @author Andrew Jarombek
 * @since 8/10/2021
 */

import java.util.HashMap;
import java.util.Map;

public class ContainsNearbyDuplicate {
    public static boolean containsNearbyDuplicate(int[] nums, int k) {
        Map<Integer, Integer> found = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            Integer j = found.get(nums[i]);
            if (j != null && Math.abs(i - j) <= k) {
                return true;
            } else {
                found.put(nums[i], i);
            }
        }

        return false;
    }

    public static void main(String[] args) {
        assert containsNearbyDuplicate(new int[] {1, 2, 3, 1}, 3);
    }
}

File: 2023/04/01/cpp/containsNearbyDuplicate.cpp
// https://leetcode.com/problems/contains-duplicate-ii/
// Author: Andrew Jarombek
// Date: 4/1/2023

#include <cassert>
#include <vector>
#include <unordered_map>

using namespace std;

bool containsNearbyDuplicate(vector<int>& nums, int k) {
    unordered_map<int, int> map {};

    for (int i = 1; i <= nums.size(); i++) {
        int j = map[nums[i - 1]];

        if (j > 0 && i - j <= k) {
            return true;
        }

        map[nums[i - 1]] = i;
    }

    return false;
}

int main() {
    vector<int> v {1, 0, 1, 1};
    assert(containsNearbyDuplicate(v, 1));

    vector<int> v2 {1, 2, 3, 1, 2, 3};
    assert(!containsNearbyDuplicate(v2, 2));
}